WaveFT method added into tuners

Bilican opened the pull request:
Initial request for #2552

BenjaminBossan commented:
Thanks for the PR. I just did a quick skim and saw that the majority of added files are inside of the `PyTorch-Wavelet-Toolbox-Custom` directory. Are these files needed? We need to organize this change to fit the structure of PEFT, so if there is useful stuff in there like tests, examples, etc., could you please re-arrange them accordingly? As an example, please check the organization of PR: https://github.com/huggingface/peft/pull/2172.

Bilican commented:
Hi @BenjaminBossan , sorry for the late update — I was caught up with finals at university. I've now removed all the unnecessary parts from the PyTorch-Wavelet-Toolbox-Custom directory as suggested. Let me know if there's anything else you'd like me to adjust!

BenjaminBossan commented on src/peft/tuners/waveft/__init__.py:
```diff
@@ -0,0 +1,20 @@
+from .model import WaveFTModel
+
+
+__all__ = ["WaveFTConfig", "WaveFTLayer", "WaveFTLinear", "WaveFTModel"]
```
Here we also need to call `register_peft_method`, see e.g. how it's done here:

https://github.com/huggingface/peft/blob/d6dbbc919576d92c31d00994e41cb503ff132c94/src/peft/tuners/fourierft/__init__.py#L24

BenjaminBossan reviewed (changes requested):
Thank you for the updates. There is, however, still a bunch of code inside `PyTorch-Wavelet-Toolbox-Custom`. This doesn't work, everything that's required for users who install PEFT must live inside of `src/peft`. I added a comment to that effect. Please check for all these files, functions, and classes in `PyTorch-Wavelet-Toolbox-Custom/` if they're required for WaveFT to work. If not, delete them. If they are, move them to `src/peft/tuners/waveft/<file>` with descriptive file names. From there, you can import those functions.

Please check the PR #2577, which is a recent PR that demonstrates how the code should be organized.

Bilican commented:
Hi @BenjaminBossan, I’ve implemented the changes you pointed out. Please let me know if there’s anything else you’d like me to revise or improve.

BenjaminBossan commented on src/peft/tuners/waveft/waverec2.py:
Why is the file named `waverec2.py` and not `waverec.py`?

BenjaminBossan commented on src/peft/tuners/waveft/waverec2.py:
```diff
@@ -0,0 +1,286 @@
+import torch
+from typing_extensions import ParamSpec, TypeVar, TypeAlias, Unpack
+
+# --- From constants.py ---
```
If `constants.py` is referring to a file in another repo, please add a permalink to that file instead, otherwise it's hard to know what you're referring to.

BenjaminBossan commented on src/peft/__init__.py:
```diff
@@ -95,6 +95,8 @@
     VeraConfig,
     VeraModel,
+    WaveFTConfig,
+    WaveFTModel,
```
We also need an import `tuners/__init__.py`.

BenjaminBossan commented on src/peft/tuners/waveft/waverec2.py:
```diff
@@ -0,0 +1,286 @@
+from typing import Any, List, Literal, NamedTuple, Optional, Protocol, Union, cast, overload
+
+import numpy as np
+import pywt
```
This package is not a PEFT dependency, so this import will fail. I don't know what package exactly that is, maybe [this one](https://pypi.org/project/waveletnn/), but we very rarely add new external dependencies, only if absolutely necessary, so I think we should remove it.

BenjaminBossan reviewed (changes requested):
Thanks a lot for cleaning up, this looks much better now.

For now, I only did a quick review, thus haven't checked the details. I tried to make WaveFT run but got into multiple issues, which I have commented on below.

I think a good next step would be to add WaveFT to our test suite and ensure that it passes. This way, we know that the basics are working. The easiest way to achieve this is to go [here](https://github.com/huggingface/peft/blob/e34852f7b67d51ba7ef871051b1236e9558c650e/tests/test_custom_models.py#L539) and add something like:

```
    ##########
    # WaveFT #
    ##########
    (
        "Vanilla MLP 1 WaveRec", "MLP", WaveFTConfig, {"target_modules": "lin0", "n_frequency": 8},
    )
```

After this, run the tests with `pytest tests/test_custom_models.py -k waveft` and check that the tests are passing.

Also, please make sure that for each new file, please ensure that the copyright notice is there and that it has the current year. Finally, once you made your changes, run `make style` to ensure that the linter is happ.y

Bilican commented:
Hi @BenjaminBossan, I’ve implemented the necessary changes (removed dependency, fixed naming convention, added tests, and fixed copyright notices) and ran the tests which seem to work without problem. Please let me know if there’s anything else you’d like me to revise or improve.

BenjaminBossan commented:
Thanks @Bilican. There have been some changes on the main branch which cause merge conflicts, but they should be easy to resolve. Could you please merge with/rebase on the latest main branch and resolve the conflicts?

BenjaminBossan commented on tests/test_custom_models.py:
Thanks for adding the tests, they are passing for me. Let's now also add the task-specific tests:

- `test_decoder_models.py`
- `test_encoder_decoder_models.py`
- `test_feature_extraction_models.py`
- `test_seq_classifier.py`

If WaveFT is also supposed to support conv layers, please also add tests for:

- `test_stablediffusion.py`
- `test_vision_models.py`
- extend `test_custom_models.py` to include conv models

Bilican commented on src/peft/tuners/waveft/layer.py:
The standard for the dreambooth task is only finetuning the attention layers thus it doesn't require support for convolutional layers. The proposed method doesn't include support for convolutional layers as they already contain much less parameters compared to linear layers.

BenjaminBossan replied:
Got it, thanks.

Bilican commented on src/peft/tuners/waveft/config.py:
```diff
@@ -0,0 +1,212 @@
+            If `False`, coefficients are initialized from a standard normal distribution scaled by a small factor.
+        proportional_parameters (`bool`):
+            If `True`, `n_frequency` is allocated proportionally to each layer's
+            `input_dim * output_dim`. Default is `False`.
```
We have not identified any scenarios where this configuration would be beneficial. However, for the sake of completeness and to allow researchers to reproduce the results presented in the paper, we have included this option. This inclusion was made purely for experimental thoroughness rather than practical utility.

Bilican commented:
Hi @BenjaminBossan, I’ve implemented the necessary changes hopefully without missing any of the points you have mentioned and ran the tests which seem to work without problem. Please let me know if there’s anything else you’d like me to revise or improve.

BenjaminBossan commented on src/peft/tuners/waveft/constants.py:
```diff
@@ -0,0 +1,24 @@
+"""
```
Let's add the copyright notice.

Bilican commented:
Sorry for the late reply @BenjaminBossan, I was away for my wedding. I do intend to continue working on this PR promptly.

BenjaminBossan commented:
No worries about the delay. And congratulations on your wedding.

Bilican commented:
Hello @BenjaminBossan . Firstly want to thank you for your kind words and patience. I couldn't replicate the CUDA errors for those scripts. They pass in my case and I have my environment details below. 

**Environment:**
- PyTorch: 2.7.0+cu126
- CUDA: 12.6  
- Python: 3.13.3
- Platform: linux

I have adressed all of the missing parts hopefully everything is as you requested. Please let me know if there’s anything else you’d like me to revise or improve.

BenjaminBossan reviewed (changes requested):
Thanks for the latest updates, we're progressing quite well and not much is still required. There is a bit of a bigger task related to deduplicating code, but it's mostly about deleting some methods, so it should hopefully not be too much work.

When you're finished with your changes, please remember to call `make style`.

> I couldn't replicate the CUDA errors for those scripts.

I tried again and the errors are gone, not sure why these tests failed earlier for me.

BenjaminBossan commented on examples/waveft_finetuning/waveft_finetuning.py:
```diff
@@ -0,0 +1,195 @@
+    # Set seed
+    if seed is not None:
+        set_seed(seed)
+    model_kwargs = {"torch_dtype": getattr(torch, torch_dtype), "device_map": device_map}
```
```suggestion
    model_kwargs = {"dtype": getattr(torch, torch_dtype), "device_map": device_map}
```

BenjaminBossan commented on examples/waveft_finetuning/waveft_finetuning.py:
I tested the example on single GPU and on two GPUs with FSDP and it worked.

BenjaminBossan reviewed (changes requested):
Thanks for cleaning up the branch and the latest updates, the PR is almost good to go. I had a few more comments, but nothing major. Please also run `make style` before committing your changes.

Bilican commented:
Hello @BenjaminBossan , hopefully this PR is good to go. Please let me know if there’s anything else you’d like me to revise or improve.

BenjaminBossan reviewed (approved):
Thanks a lot for the latest changes, the PR looks great now and I have no further comments. Again, thanks for your patience, good work.

PS: Failing CI tests are unrelated.

BenjaminBossan commented:
@Bilican Before merging, I wanted to make a run on our [MetaMath benchmark](https://github.com/huggingface/peft/tree/main/method_comparison/MetaMathQA) using the included experiment config for WaveFT. Unfortunately, both times I tried, I got an OOM error on my 24GB GPU (somewhere between 3500 and 3750 steps). Here is a plot of the sampled memory:

<img width="596" height="448" alt="image" src="https://github.com/user-attachments/assets/92249966-cdd4-4177-8412-ebab700defee" />

It doesn't look like a memory leak, rather just spiky behavior and probably this particular step contained especially long sequences or something like that.

Overall, this is not a huge deal and we can still merge. But before that, I just wanted to share this with you, maybe you have some idea for how to improve the memory situation.

Bilican commented:
Hey @BenjaminBossan, thanks so much for your patience throughout this process!
I’m really happy to see this PR finally merged. I genuinely enjoyed working with you — your comments were super thoughtful and helped me improve a lot. Appreciate it!

I’m not entirely sure what’s causing the memory usage spike, since I usually run my tests on 40GB or 80GB GPUs, I haven’t encountered this issue myself. It’s definitely interesting. I don’t think there’s a quick fix right now, but if I find anything that could improve the situation, I’ll open a new pull request.
